promql/parser/prettier.go

```diff
@@ -0,0 +1,543 @@
+				result += p.pd.newLine()
+			} else if !p.pd.labelsSplittable {
+				result += " "
+			}
+		case COMMENT:
+			if result[len(result)-1] != ' ' {
+				result += " "
+			}
+			if p.pd.isNewLineApplied {
+				result += p.pd.pad(nodeInfo.previousIndent()) + item.Val
+			} else {
+				result += item.Val
+			}
+			p.pd.isPreviousItemComment = true
+		}
+	}
+	if isNodeType(root.node, binaryExpr) && !root.hasScalar {
+		return result, nil
+	} else if isNodeType(root.node, parenExpr) && isChildOfTypeBinary(root.node) {
+		return result, nil
+	}
+	return strings.TrimSpace(result), nil
+}
+
+// refreshLexItems refreshes the contents of the lex slice and the properties
+// within it. This is expected to be called after sorting the lexItems in the
+// pre-format checks.
+func (p *prettier) refreshLexItems(items []Item) []Item {
+	return p.lexItems(stringifyItems(items))
+}
+
+// lexItems converts the given expression into a slice of Items.
+func (p *prettier) lexItems(expression string) (items []Item) {
+	var (
+		l    = Lex(expression)
+		item Item
+	)
+	for l.NextItem(&item); item.Typ != EOF; l.NextItem(&item) {
+		items = append(items, item)
+	}
+	return
+}
+
+func (p *prettier) parseExpr(expression string) error {
+	expr, err := ParseExpr(expression)
+	if err != nil {
+		return errors.Wrap(err, "parse error")
+	}
+	p.Node = expr
+	return nil
+}
+
+// isMetricNameAtomic returns true if the metric name is singular in nature.
+// This is used during sorting of lex items.
+func isMetricNameNotAtomic(metricName string) bool {
+	// We aim to convert __name__ to an ideal metric_name if the value of that labels is atomic.
+	// Since a non-atomic metric_name will contain alphabets other than a-z and A-Z including _,
+	// anything that violates this ceases the formatting of that particular label item.
+	// If this is not done then the output from the prettier might be an un-parsable expression.
+	if regexp.MustCompile("^[a-z_A-Z]+$").MatchString(metricName) && metricName != "bool" {
+		return false
+	}
+	return true
+}
+
+// skipComments advances the index until a non-comment item is
+// encountered.
+func skipComments(items []Item, index int) int {
+	var i int
+	for i = index; i < len(items); i++ {
+		if items[i].Typ == COMMENT {
+			continue
+		}
+		break
+	}
+	return i
+}
+
+// stringifyItems returns a standardized string expression
+// from slice of items without any trailing whitespaces.
+func stringifyItems(items []Item) string {
+	var expression string
+	for _, item := range items {
+		if item.Typ.IsAggregator() {
+			expression += item.Val
+			continue
+		}
+		switch item.Typ {
+		case EQL, EQL_REGEX, NEQ, NEQ_REGEX, COLON, LEFT_BRACE, LEFT_BRACKET, LEFT_PAREN,
+			RIGHT_BRACE, RIGHT_BRACKET, RIGHT_PAREN, STRING, NUMBER, IDENTIFIER, METRIC_IDENTIFIER:
+			expression += item.Val
+		case COMMA:
+			expression += item.Val + " "
+		case COMMENT:
+			expression += item.Val + "\n"
+		default:
+			expression += " " + item.Val + " "
```

Comment:
And you can remove the `if` condition above
```suggestion
		case COMMENT:
			expression += item.Val + "\n"
		default:
			expression += item.Val
```

Reply:
```
default:
			expression += item.Val
```

has to be 

```
default:
			expression += item.Val + " "
```
followed by trim of space, else things like `a + bool b` will become `a+boolb` which is unparsable.